### PR TITLE
Base Effectful.Exception on Control.Exception

### LINKED ISSUE #254: Effectful.Exception is subtly broken
Turns out that simply re-exporting functions from `safe-exception` was a mistake.

1. Functions like ~~`finally`~~ `onException` are subtly broken when used in transformers like `ExceptT` (it's highly unlikely someone will use `Effectful.Exception` to do this, but it's a possibility and I don't like that, gotta specialize them to the `Eff` monad).

2. `catch` from safe-exceptions behaves differently than `catch` from the `MonadCatch (Eff es)` instance :confused: 

I think the fix is to specialize functions to `Eff` for (1) and to leave `catch` with original semantics of the `exceptions` package, but provide helper functions like `catchAnySync` that catches all sync exceptions for (2) because this is the most common scenario where async exceptions are caught and swallowed by accident.

@ocharles @mmhat FYI since you participated in #248.

### PR DESCRIPTION
**Pull request by arybczak**
Fixes #254.

### CODE REVIEW

File: effectful-core/src/Effectful/Exception.hs [hunk 1]
```diff
@@ -1,87 +1,112 @@
...
--- Why @safe-exceptions@ and not @exceptions@? Because the former makes it much
--- easier to correctly deal with asynchronous exceptions (for more information
--- see its [README](https://github.com/fpco/safe-exceptions#readme)) and
--- provides more convenience functions.
+{-# LANGUAGE CPP #-}
+-- | Support for runtime exceptions.
+--
+-- The 'Eff' monad provides instances of 'C.MonadThrow', 'C.MonadCatch' and
+-- 'C.MonadMask', so any function that works with them can be used
+-- seamlessly. This includes API of the following libraries:
```
Comment (mmhat) [inline effectful-core/src/Effectful/Exception.hs:6]:
```suggestion
-- seamlessly. This modules provides an API similar to those of the following libraries:
```

"includes" might be a bit to strong here; Maybe we should go with a phrasing that is a bit more fuzzier.

File: effectful-core/src/Effectful/Exception.hs [hunk 2]
```diff
@@ -113,18 +138,436 @@ module Effectful.Exception
...
+-- :}
+--
+-- >>> isSyncException AsyncEx
+-- Nothing
+--
+-- >>> isAsyncException AsyncEx
+-- Just AsyncEx
+
+-- | Return 'Just' if the given exception is considered synchronous.
+isSyncException :: E.Exception e => e -> Maybe e
```
Comment (mmhat) [inline effectful-core/src/Effectful/Exception.hs:528]: Hm, shouldn't this be called `maybeSyncException`? I usually expect functions prefixed with "is" to be a predicate; I.e. to return a `Bool`.

File: effectful-core/src/Effectful/Exception.hs [hunk 3]
```diff
@@ -113,18 +138,436 @@ module Effectful.Exception
...
+isSyncException e = case E.fromException (E.toException e) of
+  Just E.SomeAsyncException{} -> Nothing
+  Nothing -> Just e
+
+-- | Return 'Just' if the given exception is considered asynchronous.
+isAsyncException :: E.Exception e => e -> Maybe e
```
Comment (mmhat) [inline effectful-core/src/Effectful/Exception.hs:534]: Same here; See the comment on `isSyncException`.

**Review by mmhat — COMMENTED** [review]
LGTM; I just left two small comments, but nothing big :+1:

File: effectful-core/src/Effectful/Exception.hs (hunk 2 above)
Comment (ocharles) [reply effectful-core/src/Effectful/Exception.hs:528]: I agree, though I'd go with `toSyncException`

Comment (mmhat) [reply effectful-core/src/Effectful/Exception.hs:528]: That's even better :+1:

Comment (arybczak) [reply effectful-core/src/Effectful/Exception.hs]: I went with option 3 - made `isSyncException` and `isAsyncException` the same type as in `safe-exceptions` and used `catchIf`/`tryIf` for `catchSync`/`trySync` ;) I missed these helpers earlier, that's why I adjusted these functions to fit the `*Just` variant, but that's unnecessary now.

File: effectful-core/src/Effectful/Exception.hs (hunk 1 above)
Comment (arybczak) [reply effectful-core/src/Effectful/Exception.hs]: I shuffled the top section, should be good now.

File: effectful-core/src/Effectful/Exception.hs [hunk 4]
```diff
@@ -1,87 +1,112 @@
...
+-- | Support for runtime exceptions.
+--
+-- This module provides thin wrappers over functions from "Control.Exception" as
+-- well as several utility functions for convenience.
+--
+-- However, the 'Eff' monad provides instances for 'C.MonadThrow',
+-- 'C.MonadCatch' and 'C.MonadMask', so you can also use the following
+-- libraries:
+--
+-- - [@exceptions@](https://hackage.haskell.org/package/exceptions)
```
Comment (Kleidukos) [inline effectful-core/src/Effectful/Exception.hs:11]: While I see the point of using `safe-` and `annotated-` variants, is there a concrete recommendation that the documentation could make for new adopters of `effectful`? Like:

> Start with using the functions of this module, and if <conditions> then switch to `exceptions`.

What would motivate someone to use the implementations in `exceptions`?

Comment (arybczak) [reply effectful-core/src/Effectful/Exception.hs]: Yeah, I removed this, see https://github.com/haskell-effectful/effectful/pull/257#discussion_r1802865352

Comment (Kleidukos) [reply effectful-core/src/Effectful/Exception.hs]: Okay, I see!

**Comment by arybczak** [comment]
Thanks everyone for the comments :slightly_smiling_face:

**Pull request merged**